### ISSUE TITLE
WIP: [RFC] Provide a secure runtime for Development Tools

### ISSUE THREAD

**Issue by raskyld** [wasmCloud/wasmCloud#1994]
### RFC Title

RFC: Provide a secure runtime for *Development Tools*

### Introduction

This RFC study whether wasmCloud could provide developers with a secure runtime for *Development Tools*.

*Development Tools* are programs used by developers throughout the software development lifecycle, it includes things like:

- Linters,
- CLIs,
- Compilers,
- Documentation auto-generators,
- Automations,
- Small local servers to serve their apps..

### Motivation

The main motivations are:

- 🔒 To help users securing their [Software Supply Chain][slsa] at the beginning of the chain (the *Producers*).
- 🌳 To foster an ecosystem of tools that leverage the platform-agnosticism provided by Webassembly and WASI.
- 🚀 To make wasmCloud a good candidate for a distributed and wasm-based build system (e.g. a distributed CI/CD platform where every step is a WASI Component), the tools used for local development could easily be run remotely!

In today's world, there are really few mechanism that prevents *Development Tools* to do nasty things. They tend to have too much capabilities on the developers' workstation. Yet, stories of legitimate package getting compromised are frequent, and their potential impact severe.

[The Deno Runtime's Security Model][deno-security] is an example of what should have been done from the beginning by our industry as a counter-measure to this attack vector. Sadly, Deno is not a language-agnostic runtime which may slow down its adoption.

*Wasmtime* and the Component Model are really good candidates for a runtime that brings security, interoperability and platform-agnosticism to *Development Tools*.

wasmCloud can be seen as a way to "clusterize" *Wasmtime*. As such, it has the potential to provide a platform for developers to build and run their [*DevOps Pipeline*][devops-pipeline]. There seems to be a demand for solutions allowing to use the same tools whether running locally or remotely in a CI/CD system, for instance:

- https://earthly.dev
- https://dagger.io

[slsa]: https://slsa.dev/spec/v1.0/terminology#software-supply-chain
[deno-security]: https://docs.deno.com/runtime/manual/basics/permissions
[devops-pipeline]: https://resources.github.com/devops/pipeline/

### Detailed Design

TODO

### Backwards Compatibility

TODO

### Alternatives Considered

TODO

### Unresolved Questions

TODO

### Conclusion

TODO

**Comment by raskyld** [comment]
For reference, @thomastaylor312 already embed wasmtime in `wash-cli` for the plugin system, see https://github.com/wasmCloud/wasmCloud/pull/1989.

**Comment by raskyld** [comment]
Also, I found an in-progress crate of wasmtime host implementations for the `wasi:http` interfaces: https://github.com/bytecodealliance/wasmtime/blob/main/crates/wasi-http/src/lib.rs

We may decide to contribute to this crate directly to make everyone benefit from our work 😄

**Comment by thomastaylor312** [comment]
Yep I use the http client from that crate to allow wash plugins to make http calls
